Handle ```$type``` macro parameter

<!--
Thank you for filing a bug report! 🐛 Please provide a short summary of the bug,
along with any information you feel relevant to replicating the bug.
-->

I tried this code:

```rust
macro_rules! foo {($type:ident) => {
    let $type = 12;
}} 

pub fn foo() {
    foo!(_a);
}
```

I expected to see this happen: no error

Instead, this happened:

```
test.rs:2:10: error: unexpected token in macro transcribe: expected ‘(’ or identifier after ‘$’, got ‘type’
    2 |     let $type = 12;
      |          ^~~~
```

### Meta

git: ```bb1e40da864addfc2d8dfc310928b7a4e7b13185```

--- a/gcc/rust/expand/rust-macro-substitute-ctx.cc
+++ b/gcc/rust/expand/rust-macro-substitute-ctx.cc
@@ -58,7 +58,8 @@ SubstituteCtx::check_repetition_amount (size_t pattern_start,
       if (macro.at (i)->get_id () == DOLLAR_SIGN)
 	{
 	  auto &frag_token = macro.at (i + 1);
-	  if (frag_token->get_id () == IDENTIFIER)
+	  if (token_id_is_keyword (frag_token->get_id ())
+	      || frag_token->get_id () == IDENTIFIER)
 	    {
 	      auto it = fragments.find (frag_token->get_str ());
 	      if (it == fragments.end ())
@@ -199,11 +200,21 @@ std::pair<std::vector<std::unique_ptr<AST::Token>>, size_t>
 SubstituteCtx::substitute_token (size_t token_idx)
 {
   auto &token = macro.at (token_idx);
+
   switch (token->get_id ())
     {
-    case IDENTIFIER:
-      rust_debug ("expanding metavar: %s", token->get_str ().c_str ());
-      return {substitute_metavar (token), 1};
+    default:
+      if (token_id_is_keyword (token->get_id ()))
+	{
+	case IDENTIFIER:
+	  rust_debug ("expanding metavar: %s", token->get_str ().c_str ());
+	  return {substitute_metavar (token), 1};
+	}
+      rust_error_at (token->get_locus (),
+		     "unexpected token in macro transcribe: expected "
+		     "%<(%> or identifier after %<$%>, got %<%s%>",
+		     get_token_description (token->get_id ()));
+      break;
       case LEFT_PAREN: {
 	// We need to parse up until the closing delimiter and expand this
 	// fragment->n times.
@@ -279,11 +290,6 @@ SubstituteCtx::substitute_token (size_t token_idx)
       // with no associated fragment and paste the dollar sign in the
       // transcription. Unsure how to do that since we always have at
       // least the closing curly brace after an empty $...
-    default:
-      rust_error_at (token->get_locus (),
-		     "unexpected token in macro transcribe: expected "
-		     "%<(%> or identifier after %<$%>, got %<%s%>",
-		     get_token_description (token->get_id ()));
     }
 
   // FIXME: gcc_unreachable() error case?

--- a/gcc/rust/lex/rust-token.cc
+++ b/gcc/rust/lex/rust-token.cc
@@ -152,6 +152,9 @@ Token::get_type_hint_str () const
 const std::string &
 Token::get_str () const
 {
+  if (token_id_is_keyword (token_id))
+    return token_id_keyword_string (token_id);
+
   // FIXME: attempt to return null again
   // gcc_assert(str != NULL);
 

--- a/gcc/rust/parse/rust-parse-impl.h
+++ b/gcc/rust/parse/rust-parse-impl.h
@@ -2146,10 +2146,10 @@ Parser<ManagedTokenSource>::parse_macro_match_fragment ()
 
   Identifier ident = "";
   auto identifier = lexer.peek_token ();
-  if (identifier->has_str ())
-    ident = identifier->get_str ();
+  if (identifier->get_id () == UNDERSCORE)
+    ident = "_";
   else
-    ident = std::string (token_id_to_str (identifier->get_id ()));
+    ident = identifier->get_str ();
 
   if (ident.empty ())
     {

--- a/gcc/testsuite/rust/compile/macro-issue2194.rs
+++ b/gcc/testsuite/rust/compile/macro-issue2194.rs
@@ -1,0 +1,7 @@
+macro_rules! foo {($type:ident) => {
+    let $type = 12;
+}} 
+
+pub fn foo() {
+    foo!(_a);
+}

